system/include/emscripten/thread_utils.h

```diff
@@ -0,0 +1,203 @@
+  std::condition_variable condition;
+  std::function<void(Callback)> work;
+  bool readyToWork = false;
+  bool finishedWork;
+  bool quit = false;
+  std::unique_ptr<std::function<void()>> resume;
+  std::mutex invokeMutex;
+
+  // The child will be asynchronous, and therefore we cannot rely on RAII to
+  // unlock for us, we must do it manually.
+  std::unique_lock<std::mutex> childLock;
+
+  static void* threadMain(void* arg) {
+    emscripten_async_call(threadIter, arg, 0);
+    return 0;
+  }
+
+  static void threadIter(void* arg) {
+    auto* parent = (sync_to_async*)arg;
+    if (parent->quit) {
+      pthread_exit(0);
+    }
+    // Wait until we get something to do.
+    parent->childLock.lock();
+    parent->condition.wait(parent->childLock, [&]() {
+      return parent->readyToWork;
+    });
+    auto work = parent->work;
+    parent->readyToWork = false;
+    // Allocate a resume function, and stash it on the parent.
+    parent->resume = std::make_unique<std::function<void()>>([parent, arg]() {
+      // We are called, so the work was finished. Notify the caller.
+      parent->finishedWork = true;
+      parent->childLock.unlock();
+      parent->condition.notify_one();
+      // Look for more work. Doing this asynchronously ensures that we continue
+      // after the current call stack unwinds (avoiding constantly adding to the
+      // stack, and also running any remaining code the caller had, like
+      // destructors). TODO: add an option to do a synchronous call here in some
+      // cases, which would avoid the time delay caused by a browser setTimeout.
+      emscripten_async_call(threadIter, arg, 0);
+    });
+    // Run the work function the user gave us. Give it a pointer to the resume
+    // function.
+    work(parent->resume.get());
+  }
+
+public:
+  sync_to_async() : childLock(mutex) {
+    // The child lock is associated with the mutex, which takes the lock as we
+    // connect them, and so we must free it here so that the child can use it.
+    // Only the child will lock/unlock it from now on.
+    childLock.unlock();
+
+    // Create the thread after the lock is ready.
+    thread = std::make_unique<std::thread>(threadMain, this);
+  }
+
+  ~sync_to_async() {
+    // Wake up the child to tell it to quit.
+    invoke([&](Callback func){
+      quit = true;
+      (*func)();
+    });
+
+    thread->join();
+  }
+};
+
+void sync_to_async::invoke(std::function<void(Callback)> newWork) {
+  // Use the invokeMutex to prevent more than one invoke being in flight at a
+  // time, so that this is usable from multiple threads safely.
+  std::lock_guard<std::mutex> invokeLock(invokeMutex);
+
+  // Send the work over.
+  {
+    std::lock_guard<std::mutex> lock(mutex);
+    work = newWork;
+    finishedWork = false;
+    readyToWork = true;
+  }
+  condition.notify_one();
+
+  // Wait for it to be complete.
+  std::unique_lock<std::mutex> lock(mutex);
+  condition.wait(lock, [&]() {
+    return finishedWork;
+  });
```

Comment:
Releasing the lock before notifying the worker thread makes it possible for the worker to respond before we can re-acquire the thread and wait for its response. Missing that response hangs the invoking thread.

```suggestion
  // Send the work over.
  std::lock_guard<std::mutex> lock(mutex);
  work = newWork;
  finishedWork = false;
  readyToWork = true;
  condition.notify_one();

  // Wait for it to be complete.
  condition.wait(lock, [&]() {
    return finishedWork;
  });
```

Reply:
Ah, I just saw this now after opening #15611. Looks like I still take the lock two times there, although it should at least be safe now, I think? But it probably makes sense to lock once, yes...